File: ObjC_Example/Fullboard/FullBoardAdWebViewController.m
//
//  FullBoardAdWebViewController.m
//  ObjC_Example
//
//  Copyright © 2017年 FAN Communications. All rights reserved.
//

#import "FullBoardAdWebViewController.h"
#import <WebKit/WebKit.h>
#import <NendAd/NADFullBoardLoader.h>


@interface FullBoardAdWebViewController () <UIScrollViewDelegate, NADFullBoardDelegate>

@property (nonatomic, strong) WKWebView *webView;
@property (nonatomic, strong) NADFullBoardLoader *loader;
@property (nonatomic, strong) NADFullBoard *ad;

@end

@implementation FullBoardAdWebViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    // Do any additional setup after loading the view.
    self.webView = [[WKWebView alloc] initWithFrame:self.view.frame configuration:[WKWebViewConfiguration new]];
    self.webView.translatesAutoresizingMaskIntoConstraints = NO;
    [self.view addSubview:self.webView];
    self.webView.scrollView.delegate = self;
    
    [self.view addConstraints:@[
        [NSLayoutConstraint constraintWithItem:self.webView
                                     attribute:NSLayoutAttributeTop
                                     relatedBy:NSLayoutRelationEqual
                                        toItem:self.view
                                     attribute:NSLayoutAttributeTop
                                    multiplier:1
                                      constant:0],
        [NSLayoutConstraint constraintWithItem:self.webView
                                     attribute:NSLayoutAttributeLeading
                                     relatedBy:NSLayoutRelationEqual
                                        toItem:self.view
                                     attribute:NSLayoutAttributeLeading
                                    multiplier:1
                                      constant:0],
        [NSLayoutConstraint constraintWithItem:self.webView
                                     attribute:NSLayoutAttributeTrailing
                                     relatedBy:NSLayoutRelationEqual
                                        toItem:self.view
                                     attribute:NSLayoutAttributeTrailing
                                    multiplier:1
                                      constant:0],
        [NSLayoutConstraint constraintWithItem:self.webView
                                     attribute:NSLayoutAttributeBottom
                                     relatedBy:NSLayoutRelationEqual
                                        toItem:self.view
                                     attribute:NSLayoutAttributeBottom
                                    multiplier:1
                                      constant:0],
    ]];
    
    
    [self.webView loadRequest:[NSURLRequest requestWithURL:[NSURL URLWithString:@"https://www.nend.net"]]];
    self.loader = [[NADFullBoardLoader alloc] initWithSpotID:485504 apiKey:@"30fda4b3386e793a14b27bedb4dcd29f03d638e5"];
    [self loadAd];
}

- (void)didReceiveMemoryWarning {
    [super didReceiveMemoryWarning];
    // Dispose of any resources that can be recreated.
}

- (void)loadAd
{
    __weak typeof(self) weakSelf = self;
    [self.loader loadAdWithCompletionHandler:^(NADFullBoard *ad, NADFullBoardLoaderError error) {
        if (weakSelf && ad) {
            ad.delegate = weakSelf;
            weakSelf.ad = ad;
        }
    }];
}

#pragma mark - NADFullBoardDelegate

- (void)NADFullBoardDidDismissAd:(NADFullBoard *)ad
{
    [self loadAd];
}

#pragma mark - UIScrollViewDelegate

- (void)scrollViewDidScroll:(UIScrollView *)scrollView
{
    if (scrollView.contentOffset.y >= scrollView.contentSize.height - scrollView.frame.size.height) {
        if (self.ad) {
            dispatch_after(dispatch_time(DISPATCH_TIME_NOW, (int64_t)(0.5f * NSEC_PER_SEC)), dispatch_get_main_queue(), ^{
                [self.ad showFromViewController:self];
            });
        }
    }
}

@end

File: Swift_Example/Fullboard/FullBoardAdWebViewController.swift
//
//  FullBoardAdWebViewController.swift
//  ObjC_Example
//
//  Copyright © 2017年 FAN Communications. All rights reserved.
//

import UIKit
import WebKit
import NendAd

class FullBoardAdWebViewController: UIViewController {

    private var webView: WKWebView!
    private let loader = NADFullBoardLoader(spotID: 485504, apiKey: "30fda4b3386e793a14b27bedb4dcd29f03d638e5")
    fileprivate var ad: NADFullBoard?
    
    override func viewDidLoad() {
        super.viewDidLoad()

        // Do any additional setup after loading the view.
        self.webView = WKWebView(frame: self.view.frame, configuration: WKWebViewConfiguration())
        self.webView.translatesAutoresizingMaskIntoConstraints = false
        self.view.addSubview(self.webView)
        
        view.addConstraints([
            NSLayoutConstraint.init(item: self.webView!,
                                    attribute: .top,
                                    relatedBy: .equal,
                                    toItem: self.view,
                                    attribute: .top,
                                    multiplier: 1,
                                    constant: 0),
            NSLayoutConstraint.init(item: self.webView!,
                                    attribute: .leading,
                                    relatedBy: .equal,
                                    toItem: self.view,
                                    attribute: .leading,
                                    multiplier: 1,
                                    constant: 0),
            NSLayoutConstraint.init(item: self.webView!,
                                    attribute: .trailing,
                                    relatedBy: .equal,
                                    toItem: self.view,
                                    attribute: .trailing,
                                    multiplier: 1,
                                    constant: 0),
            NSLayoutConstraint.init(item: self.webView!,
                                    attribute: .bottom,
                                    relatedBy: .equal,
                                    toItem: self.view,
                                    attribute: .bottom,
                                    multiplier: 1,
                                    constant: 0)
        ])
        
        self.webView.scrollView.delegate = self
        self.webView.load(URLRequest(url: URL(string: "https://www.nend.net")!))
        self.loadAd()
    }

    override func didReceiveMemoryWarning() {
        super.didReceiveMemoryWarning()
        // Dispose of any resources that can be recreated.
    }
    
    func loadAd() {
        self.loader!.loadAd(completionHandler: { [weak self] (ad, error) in
            guard let strongSelf = self, let fullboardAd = ad else {
                return
            }
            fullboardAd.delegate = strongSelf
            strongSelf.ad = fullboardAd
        })
    }
}

extension FullBoardAdWebViewController: NADFullBoardDelegate {
    
    func nadFullBoardDidDismissAd(_ ad: NADFullBoard!) {
        self.loadAd()
    }
}

extension FullBoardAdWebViewController: UIScrollViewDelegate {
    
    func scrollViewDidScroll(_ scrollView: UIScrollView) {
        if scrollView.contentOffset.y >= scrollView.contentSize.height - scrollView.frame.size.height {
            if let fullboardAd = self.ad {
                DispatchQueue.main.asyncAfter(deadline: .now() + 0.5) {
                    fullboardAd.show(from: self)
                }
            }
        }
    }
}
